

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Tutorial &mdash; orthofisher  documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=9edc463e" />
      <link rel="stylesheet" type="text/css" href="../_static/custom.css?v=b745eb55" />

  
    <link rel="shortcut icon" href="../_static/flavicon.png"/>
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../_static/documentation_options.js?v=5929fcd5"></script>
      <script src="../_static/doctools.js?v=fd6eb6e6"></script>
      <script src="../_static/sphinx_highlight.js?v=6ffebe34"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Change log" href="../change_log/index.html" />
    <link rel="prev" title="Install" href="../install/index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html">
            
              <img src="../_static/orthofisher.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../about/index.html">About</a></li>
<li class="toctree-l1"><a class="reference internal" href="../arguments/index.html">Arguments</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install/index.html">Install</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#creating-hmms">1) Creating HMMs</a></li>
<li class="toctree-l2"><a class="reference internal" href="#download-the-test-data">2) Download the test data</a></li>
<li class="toctree-l2"><a class="reference internal" href="#run-orthofisher">3) Run orthofisher</a></li>
<li class="toctree-l2"><a class="reference internal" href="#examine-output">4) Examine output</a></li>
<li class="toctree-l2"><a class="reference internal" href="#estimating-gene-family-copy-number">5) Estimating gene family copy number</a></li>
<li class="toctree-l2"><a class="reference internal" href="#running-orthofisher-faster">6) Running orthofisher faster</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../change_log/index.html">Change log</a></li>
<li class="toctree-l1"><a class="reference internal" href="../other_software/index.html">Other software</a></li>
<li class="toctree-l1"><a class="reference internal" href="../frequently_asked_questions/index.html">FAQ</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">orthofisher</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Tutorial</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="tutorial">
<h1>Tutorial<a class="headerlink" href="#tutorial" title="Link to this heading"></a></h1>
<p><strong>orthofisher</strong> enables researchers to conduct high-throughput identification
of orthologous genes using Hidden Markov Models (HMM). This tutorial covers the
easy-to-implement workflow needed for using orthofisher.</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
<hr class="docutils" />
<section id="creating-hmms">
<h2>1) Creating HMMs<a class="headerlink" href="#creating-hmms" title="Link to this heading"></a></h2>
<p>Before using orthofisher, users will have to create (or curate) sets of HMMs of interest.
HMMs can be curated from various sources including databases such as <a class="reference external" href="https://pfam.xfam.org/">Pfam database</a>. In this tutorial, we will go over creating your own HMMs.</p>
<p>First, we have to decide on genes of interest, download them, and align them. In this tutorial,
we will download protein sequences that are top hits in BLASTP searches
to the GAL1, GAL7, and GAL10 genes in <em>Saccharomyces cerevisiae</em>, which make up a very famous
metabolic gene cluster in fungi. To do so, I individually BLASTed the
<a class="reference external" href="https://www.yeastgenome.org/locus/S000000224/protein">GAL1</a>,
<a class="reference external" href="https://www.yeastgenome.org/locus/S000000222/protein">GAL7</a>, and
<a class="reference external" href="https://www.yeastgenome.org/locus/S000000223/protein">GAL10</a> genes against NCBI’s default database.
Next, I downloaded the aligned FASTA sequences for each gene, which can be downloaded here:</p>
<p class="centered">
<strong>Download GAL1/7/10 alignments:
<a class="reference download internal" download="" href="../_downloads/5f807891cd2926fbcc3bf875dfa07413/gal_sequences.tar.gz"><code class="xref download docutils literal notranslate"><span class="pre">GAL</span> <span class="pre">alignments</span></code></a></strong></p><p>Unzip the directory and from within the new directory, and generate alignments for each set
of protein sequences. To do so, I will use <a class="reference external" href="http://www.clustal.org/omega/">Clustal Omega</a>
alignment protocol, which is available via <a class="reference external" href="https://www.ebi.ac.uk/Tools/msa/clustalo/">EMBL-EBI</a>.
However, other alignment software like <a class="reference external" href="https://mafft.cbrc.jp/alignment/software/">MAFFT</a> can be used.
Next, use the resulting alignments as input into the hmmbuild function from the HMMER suite
using the following commands:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$<span class="w"> </span>hmmbuild<span class="w"> </span>Gal1p.hmm<span class="w"> </span>Gal1p.aln.faa
$<span class="w"> </span>hmmbuild<span class="w"> </span>Gal7p.hmm<span class="w"> </span>Gal7p.aln.faa
$<span class="w"> </span>hmmbuild<span class="w"> </span>Gal10p.hmm<span class="w"> </span>Gal10p.aln.faa
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="download-the-test-data">
<h2>2) Download the test data<a class="headerlink" href="#download-the-test-data" title="Link to this heading"></a></h2>
<p>For ease of use, this tutorial will rely on a small dataset, which can be downloaded using
the following link:</p>
<p class="centered">
<strong>Download test data:
<a class="reference download internal" download="" href="../_downloads/975936890e1722edb688af9d88632067/orthofisher_tutorial.tar.gz"><code class="xref download docutils literal notranslate"><span class="pre">tutorial</span> <span class="pre">dataset</span></code></a></strong></p><p>Next, unzip the downloaded directory and change directory to the newly downloaded directory.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$<span class="w"> </span><span class="nb">cd</span><span class="w"> </span>path_to_unzipped_directory/orthofisher_tutorial
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="run-orthofisher">
<h2>3) Run orthofisher<a class="headerlink" href="#run-orthofisher" title="Link to this heading"></a></h2>
<p>Two arguments are required when using orthofisher.</p>
<p>The first arument, -f/–fasta, points to a two column tab delimited file that specifies
the location of fasta files that will be searched using HMMs. Typically, these are protein
fasta files from the entire genome/transcriptome of an organism. Additionally, the second
column of the file specifies the identifier for the organism. This will be used when
representing sequences from a given proteome in a multi-fasta file. In this tutorial,
this is file <strong>fasta_arg.txt</strong>.</p>
<p>The second argument, -m is a file that points to the location of HMMs that you wish to
identify or <em>fish out</em> of a given proteome. In this tutorial, this is file <strong>hmms.txt</strong>.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$<span class="w"> </span>orthofisher<span class="w"> </span>-m<span class="w"> </span>hmms.txt<span class="w"> </span>-f<span class="w"> </span>fasta_arg.txt
</pre></div>
</div>
<p>If you rerun the same command to the same output directory, include <code class="docutils literal notranslate"><span class="pre">--force</span></code>:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$<span class="w"> </span>orthofisher<span class="w"> </span>-m<span class="w"> </span>hmms.txt<span class="w"> </span>-f<span class="w"> </span>fasta_arg.txt<span class="w"> </span>--force
</pre></div>
</div>
<p>To write larger raw outputs (<code class="docutils literal notranslate"><span class="pre">all_sequences/</span></code> and <code class="docutils literal notranslate"><span class="pre">hmmsearch_output/</span></code>), include
<code class="docutils literal notranslate"><span class="pre">--verbose-output</span></code>:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$<span class="w"> </span>orthofisher<span class="w"> </span>-m<span class="w"> </span>hmms.txt<span class="w"> </span>-f<span class="w"> </span>fasta_arg.txt<span class="w"> </span>--verbose-output
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="examine-output">
<h2>4) Examine output<a class="headerlink" href="#examine-output" title="Link to this heading"></a></h2>
<p>In the current working directory, a subdirectory will be made titled <em>orthofisher_output</em>.
By default, orthofisher writes a slim output with:</p>
<ol class="arabic simple">
<li><p>scog: a directory of single copy orthologous HMMs identified in the various fasta files</p></li>
<li><p>long_summary.txt</p></li>
<li><p>short_summary.txt</p></li>
</ol>
<p>If you use <code class="docutils literal notranslate"><span class="pre">--verbose-output</span></code>, orthofisher also writes:</p>
<ol class="arabic simple">
<li><p>all_sequences: multi-fasta file sequences of every hit identified during sequence similarity search.</p></li>
<li><p>hmmsearch_output: output files generated during hmmsearches.</p></li>
</ol>
<p>Also, two text files are made with helpful information that summarizes all the searches:</p>
<p>1. long_summary.txt: Hits identified during sequence similarity search per fasta file per HMM.
HMMs are considered single-copy, multi-copy, or absent in a given fasta file.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$<span class="w"> </span>cat<span class="w"> </span>orthofisher_output/long_summary.txt
Clavispora_lusitaniae_P5.faa<span class="w"> </span>Gal10p.hmm<span class="w"> </span>single-copy<span class="w"> </span><span class="m">1</span><span class="w"> </span>QFZ46524.1<span class="w"> </span><span class="m">2</span>.8e-180<span class="w"> </span><span class="m">597</span>.0
Saccharomyces_cerevisiae_S288C.faa<span class="w"> </span>Gal1p.hmm<span class="w"> </span>multi-copy<span class="w"> </span><span class="m">2</span><span class="w"> </span>NP_009576.1<span class="w"> </span><span class="m">3</span>.5e-250<span class="w"> </span><span class="m">830</span>.2
small_set.faa<span class="w"> </span>Gal7p.hmm<span class="w"> </span>absent<span class="w"> </span><span class="m">0</span><span class="w"> </span>NA<span class="w"> </span>NA<span class="w"> </span>NA
</pre></div>
</div>
<p>col. 1: Query proteome fasta file. <br />
col. 2: HMM file used during sequence similarity search. <br />
col. 3: The sequence represented by the HMM is considered single_copy, multi-copy, or absent in a query proteome. <br />
col. 4: Absolute copy number of hits from the sequence similarity search. <br />
col. 5: The fasta entry identifier of the gene identified. <br />
col. 6: e-value for the identified hit (<code class="docutils literal notranslate"><span class="pre">NA</span></code> if absent). <br />
col. 7: bitscore for the identified hit (<code class="docutils literal notranslate"><span class="pre">NA</span></code> if absent). <br /></p>
<p>2. short_summary.txt: Summary of the absolute number and percentage of single-copy, multi-copy, or absent HMMs
per fasta file.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$<span class="w"> </span>cat<span class="w"> </span>orthofisher_output/short_summary.txt
file_name<span class="w">       </span>single-copy<span class="w">     </span>multi-copy<span class="w">      </span>absent<span class="w">  </span>per_single-copy<span class="w"> </span>per_multi-copy<span class="w">  </span>per_absent
Clavispora_lusitaniae_P5.faa<span class="w">    </span><span class="m">3</span><span class="w">       </span><span class="m">0</span><span class="w">       </span><span class="m">0</span><span class="w">       </span><span class="m">1</span>.0<span class="w">     </span><span class="m">0</span>.0<span class="w">     </span><span class="m">0</span>.0
Kluyveromyces_lactis_NRRL_Y-1140.faa<span class="w">    </span><span class="m">3</span><span class="w">       </span><span class="m">0</span><span class="w">       </span><span class="m">0</span><span class="w">       </span><span class="m">1</span>.0<span class="w">     </span><span class="m">0</span>.0<span class="w">     </span><span class="m">0</span>.0
Saccharomyces_cerevisiae_S288C.faa<span class="w">      </span><span class="m">2</span><span class="w">       </span><span class="m">1</span><span class="w">       </span><span class="m">0</span><span class="w">       </span><span class="m">0</span>.67<span class="w">    </span><span class="m">0</span>.33<span class="w">    </span><span class="m">0</span>.0
Yarrowia_lipolytica_DSM3286.faa<span class="w"> </span><span class="m">3</span><span class="w">       </span><span class="m">0</span><span class="w">       </span><span class="m">0</span><span class="w">       </span><span class="m">1</span>.0<span class="w">     </span><span class="m">0</span>.0<span class="w">     </span><span class="m">0</span>.0
</pre></div>
</div>
<p>col. 1: Query proteome fasta file. <br />
col. 2-4: Absolute number of sequences represented by HMMs that are present in single-copy, multi-copy, or absent. <br />
col. 5-7: Percetange of sequences represented by HMMs that are present in single-copy, multi-copy, or absent. <br /></p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="estimating-gene-family-copy-number">
<h2>5) Estimating gene family copy number<a class="headerlink" href="#estimating-gene-family-copy-number" title="Link to this heading"></a></h2>
<p>In the original paper describing orthofisher, we note that orthofisher can also be used to estimate gene family
copy number using a zinc finger, C2H2 type, (Pfam: PF00096) as an example. Following that, we will now estimate
PF00096 domain copy number among the same set of proteomes.</p>
<p>To do so, first download the PF00096 HMM from <a class="reference external" href="http://pfam.xfam.org/family/zf-C2H2#tabview=tab6">EMBL-EBI</a>, which is also available here:</p>
<p class="centered">
<strong>Download Zinc Finger, C2H2 type, HMM:
<a class="reference download internal" download="" href="../_downloads/6d3fc0dedd82495c1739ab114c9bba21/PF00096.hmm"><code class="xref download docutils literal notranslate"><span class="pre">PF00096.hmm</span></code></a></strong></p><p>Next, follow the step described above to run orthofisher using PF00096.hmm. Of note, we recommend lowering the bitscore threshold
to a less stringent value. The default bitscore threshold is 85% of the best hit, which follows the <a class="reference external" href="https://busco.ezlab.org/">BUSCO</a> pipeline.
However, depending on the research question, lowering the threshold may be reasonable. For example, try running orthofisher using the -b parameter
set to 0.85 and 0.2. Using a threshold of 0.2, Clavispora_lusitaniae_P5.faa has 27 copies, Kluyveromyces_lactis_NRRL_Y-1140.faa has 23 copies,
Saccharomyces_cerevisiae_S288C.faa has 31 copies, and Yarrowia_lipolytica_DSM3286.faa has 28 copies of PF00096 whereas the same species have 1, 1, 1, and 3
copies, respectively, using default parameters.</p>
<p>We provide an additional level of user-flexibility by having a e-value threshold that can be used during HMM-based sequence similarity search. To change the
e-value threshold from the default value of 1e-3, use the -e parameter.</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="running-orthofisher-faster">
<h2>6) Running orthofisher faster<a class="headerlink" href="#running-orthofisher-faster" title="Link to this heading"></a></h2>
<p>In this section, I go over some suggestions to make orthofisher run faster.</p>
<p>As of version 1.0.0,
two new arguments have been added to orthofisher. The first is -c/--cpu, which can be use for
multithreading during the HMMsearch. The second is -o/--output_dir, which can be used to specify
the name of the output file. These arguments enable users to accelerate the HMMsearch and
pseudo-parallelize orthofisher runs.</p>
<p>To pseudo-parallelize orthofisher runs, users can split up their -f/--fasta input file into
multiple pieces and then run orthofisher using each smaller file in an array of jobs.
When doing so, users can write to different output directories for each input file and then
combine results at the end.</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../install/index.html" class="btn btn-neutral float-left" title="Install" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../change_log/index.html" class="btn btn-neutral float-right" title="Change log" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2020 Jacob L. Steenwyk.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>
    <!-- Theme Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-104875636-1"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());

      gtag('config', 'UA-104875636-1', {
          'anonymize_ip': false,
      });
    </script> 

</body>
</html>